LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DefaultLoRaDataRequestHandler.cs

```diff
@@ -156,30 +177,30 @@ public async Task<LoRaDeviceRequestProcessResult> ProcessRequestAsync(LoRaReques
                         loRaDevice.UpdateRegion(request.Region.LoRaRegion, acceptChanges: false);
                 }
 
+                #region FrameCounterDown
                 // if deduplication already processed the next framecounter down, use that
                 var fcntDown = loRaADRResult?.FCntDown != null ? loRaADRResult.FCntDown : bundlerResult?.NextFCntDown;
+                LogNotNullFrameCounterDownState(loRaDevice, fcntDown);
 
-                if (fcntDown.HasValue)
-                {
-                    LogFrameCounterDownState(loRaDevice, fcntDown.Value);
-                }
-
-                // If it is confirmed it require us to update the frame counter down
+                // If we can send message downstream, we need to update the frame counter down
                 // Multiple gateways: in redis, otherwise in device twin
-                if (requiresConfirmation)
+                if (requiresConfirmation && !skipDownstreamToAvoidCollisions)
                 {
                     fcntDown = await EnsureHasFcntDownAsync(loRaDevice, fcntDown, payloadFcntAdjusted, frameCounterStrategy);
 
                     // Failed to update the fcnt down
-                    // In multi gateway scenarios it means the another gateway was faster than using, can stop now
+                    // In multi gateway scenarios it means another gateway has won the race to handle this message, can stop now
                     if (fcntDown <= 0)
                     {
                         return new LoRaDeviceRequestProcessResult(loRaDevice, request, LoRaDeviceRequestFailedReason.HandledByAnotherGateway);
                     }
                 }
+                #endregion
 
-                var validFcntUp = isFrameCounterFromNewlyStartedDevice || (payloadFcntAdjusted > loRaDevice.FCntUp);
-                if (validFcntUp || isConfirmedResubmit)
+                var canSendUpstream = isFrameCounterFromNewlyStartedDevice
```

Comment:
I think here we are not sending here upstream but only calling the decoders? There is another variable called sendUpstream below, what is the difference between the two?

Reply:
> I think here we are not sending here upstream but only calling the decoders?

Correct. Patrick had a similar comment, the decoder logic could be potentially moved in the downstream part. I created #1074 for this since it's not a change introduced in this PR.

Reply:
we always need the decoder for the upstream but also for the downstream. I think the decoder call, will have to happen, if either, down or upstream is required. Because of the feature that the decoder could compose a C2D message.

Reply:
sorry yes, I meant the particular logic for the C2D feature could be moved in the downstream part (in the story it's correct).

Reply:
Will resolve this now but please re-open if you have further questions @Mandur